### PR TITLE
ShlomiShitrit/issue25 - add more then one parameter query

### LINKED ISSUE #25: Add more then one parameter query

### PR DESCRIPTION
**Pull request by ShlomiShitrit**
Fixes #25

Now the query can handle multiple parameters.

For example:
if the attribute "name"  has been with the operator "startWith" twice - for "a" and for "b"
and the attribute "gender" has been chosen with the operator "startWith" twice - for "m"  and for "f":

The results are:

image
from  sw_characters csv file

### CODE REVIEW

**Comment by ShlomiShitrit** [comment]
The feature is not working, need to debug and test

**Comment by ShlomiShitrit** [comment]
> The feature is not working, need to debug and test

Was using different branch by mistake, works fine

@nadavWeisler  waiting for your review

File: BotUI/src/app/api/root/route.ts [hunk 1]
```diff
@@ -1,32 +1,29 @@
...
+import { Attribute } from "@/app/general/interfaces";
 import { getOperator } from "@/app/general/utils";
+import { strOrNum } from "@/app/general/types";
 import { NextRequest, NextResponse } from "next/server";
 import fs from "fs";
 import csvParser from "csv-parser";
 
 export async function POST(request: NextRequest) {
     const req = await request.json();
-    const queryReq = req.queryReq as QueryReq[];
```
Comment (nadavWeisler) [inline BotUI/src/app/api/root/route.ts:9]: I like this change, much more intuative

File: BotUI/src/app/api/root/route.ts [hunk 2]
```diff
@@ -1,32 +1,29 @@
...
         fs.createReadStream(filePath)
             .pipe(csvParser())
             .on("data", (row: any) => {
-                const isRequired = queryReq.every((query, index) =>
-                    operators[index](
-                        row[query.parameter],
-                        ...query.functionParams
-                    )
-                );
+                const isRequired = isRowRequired(attributes, operators, row);
```
Comment (nadavWeisler) [inline BotUI/src/app/api/root/route.ts:25]: nice

Comment (nadavWeisler) [reply BotUI/src/app/api/root/route.ts:25]: but why create param and not just add isRowRequired(attributes, operators, row) in an "if" statement

File: BotUI/src/app/components/ChatBox/ChatBox.tsx [hunk 3]
```diff
@@ -66,8 +70,8 @@ export default function ChatBox({ bot }: ChatBoxProps) {
 
     useEffect(() => {
         if (endChat.state) {
-            const params = handleEndChat();
-            setQueryParams(params);
+            const queryAttributes = handleEndChat();
```
Comment (nadavWeisler) [inline BotUI/src/app/components/ChatBox/ChatBox.tsx:73]: Why create this param and not just add it to the state?

File: BotUI/src/app/general/interfaces.ts [hunk 4]
```diff
@@ -84,27 +84,14 @@ export enum DataType {
...
-    value: string;
+export interface Attribute {
+    name: string;
     params: strOrNum[];
     operator: string;
 }
 
 export interface QueryWords {
-    [key: string]: NumericAttribute | StringAttribute | null;
+    [key: string]: Attribute;
```
Comment (nadavWeisler) [inline BotUI/src/app/general/interfaces.ts:94]: REALLY NICE!

**Review by nadavWeisler — COMMENTED** [review]
I really like your use of hooks, must admit your understanding of react framework is amazing, probobly better then mine.

File: BotUI/src/app/api/root/route.ts (hunk 2 above)
Comment (ShlomiShitrit) [reply BotUI/src/app/api/root/route.ts:25]: done

File: BotUI/src/app/components/ChatBox/ChatBox.tsx (hunk 3 above)
Comment (ShlomiShitrit) [reply BotUI/src/app/components/ChatBox/ChatBox.tsx:73]: done

**Comment by ShlomiShitrit** [comment]
> I really like your use of hooks, must admit your understanding of react framework is amazing, probobly better then mine.

Thanks!

**Comment by ShlomiShitrit** [comment]
@nadavWeisler I fixed your notes, let me know

**Review by nadavWeisler — APPROVED** [review]

**Pull request merged**